### ISSUE TITLE
Batch operations always returning 404: Not Found response

### ISSUE THREAD

**Issue by clingle** [google/google-api-php-client#1320]
Using the Google_Service_PeopleService API provided in the [google-api-php-client-services](https://github.com/google/google-api-php-client-services) library, I'm attempting to create contacts via the Google_Http_Batch class. 

I've tested executing the request on its own prior to adding it to the batch and it works correctly, however when attempting to use batching it always returns a 404 Not Found Exception as the Response.

I searched around for similar issues and it seems like this can be caused by methods not allowing batching but I don't see anywhere in the People API documentation that specifies which methods can and can't be batched.

Here is a variable dump (using Laravel's `dump()` function) for the Google_Http_Batch object:
```
Google_Http_Batch {#733 ▼
  -boundary: 350902964
  -requests: array:1 [▼
    1 => Request {#764 ▼
      -method: "POST"
      -requestTarget: null
      -uri: Uri {#766 ▼
        -scheme: "https"
        -userInfo: ""
        -host: "people.googleapis.com"
        -port: null
        -path: "/v1/people:createContact"
        -query: ""
        -fragment: ""
      }
      -headers: array:3 [▶]
      -headerNames: array:3 [▶]
      -protocol: "1.1"
      -stream: Stream {#765 ▶}
    }
  ]
  -client: Google_Client {#732 ▶}
  -rootUrl: "https://www.googleapis.com"
  -batchPath: "batch"
}
```

And here is both the request and response bodies for the attempted operation:
```
// Request Body
--350902964\n
Content-Type: application/http
Content-Transfer-Encoding: binary
MIME-Version: 1.0
Content-ID: 1

POST /v1/people:createContact HTTP/1.1
Host:people.googleapis.com
content-type:application/json
X-Php-Expected-Class:Google_Service_PeopleService_Person

{"names":[{"familyName":"Last","givenName":"First"}],"birthdays":[{"text":"1975-01-01"}],"emailAddresses":[{"type":"other","value":"firstlast@gmail.com"}],"phoneNumbers":[{"type":"other","value":"+18005555555"}]}\n
\n
--350902964--


// Response Body
--batch_pb1ntA148M4_AAcIp_VroDQ
Content-Type: application/http
Content-ID: response-1
\r\n
HTTP/1.1 404 Not Found
Content-Type: text/html; charset=UTF-8
Date: Thu, 19 Oct 2017 20:44:28 GMT
Expires: Thu, 19 Oct 2017 20:44:28 GMT
Cache-Control: private, max-age=0
Content-Length: 9

Not Found
--batch_pb1ntA148M4_AAcIp_VroDQ--
```

If this appears to be an issue with the `Google_Service_PeopleService` class instead of the `Google_Http_Batch` class, I'll be more than happy to create an issue on the [google-api-php-client-services](https://github.com/google/google-api-php-client-services) and close this one.

Any clarification on whether I'm doing anything incorrectly or if I'm even able to batch this method at all would be greatly appreciated!

**Comment by mattwhisenhunt** [comment]
It looks like there are only batch get methods: https://developers.google.com/people/api/rest/.

**Comment by clingle** [comment]
Are only methods with "batch" in the name able to be batched? I'm not knowledgeable on Google's method-naming standard.

Our use-case is to allow users to be able to sync contacts between our system and Google, so requiring a request to be made for every single creation during the initial sync of ~2k contacts or more will take quit a bit of time, even if using Guzzle in parallel. Is there a better way to approach that issue if using `Google_Http_Batch` isn't an option?

**Comment by mattwhisenhunt** [comment]
Looks like you might want to use a different API, Contacts: https://developers.google.com/google-apps/contacts/v3/#batch_operations

**Comment by clingle** [comment]
We originally were (are) using the Contacts API but we had to write our own library because the api-php-client-services repo has no classes for it, at least that I could find. Contacts appears to be on the path to deprecation because it uses Gdata, and even recommends the People API in its documentation at the top of its main page now. I guess at this point it's a "wait and see" if the People API gets more of the batch functionality that is currently present in the Contacts API. Either way, thanks for the help!

**Comment by amosyuen** [comment]
Looks like the `rootUrl` is set incorrectly. It should point to `https://people.googleapis.com` instead of `https://www.googleapis.com`

**Comment by clingle** [comment]
I was passing the client as follows:
```php
$this->client->setAccessToken($accessToken);
$this->client->setUseBatch(true);
$batch = new Google_Http_Batch($this->client);
```
But you mentioning the wrong `rootUrl` turned out to be helpful. The third parameter of the `Google_Http_Batch` constructor accepts a `rootUrl` to use, so the new instance looks like this:
```php
$batch = new Google_Http_Batch($this->client, false, 'https://people.googleapis.com');
```

I'm almost positive I tried setting the `rootUrl` previously when attempting to find a solution with no success (perhaps I was incorrectly setting it for each individual request being added to the batch instead of the `Google_Http_Batch` instance itself) but it appears to be working now. The issue now is hitting the rate limit of the API itself, but I had already submitted a request to have our limit increased so I'm waiting for the results of that. Good find!
